test/debugInfo.fromasm

```diff
@@ -47,17 +46,11 @@
     (get_local $0)
    )
   )
+  ;;@ even-opted.cpp:3:0
```

Comment:
why are test results changing? should this be just a refactoring without functional changes?

Reply:
My PR description was pretty sparse on detail, and I forgot to write down at least one case:
* Because we change the timing of when we convert things to calls, optimizations happen in different orders. This lets constant folding be more effective, as seen in `big_uint_div_u` (in test/unit.fromasm), because we can constant fold the `i32.div_u` instruction directly, and then it doesn't generate a clamping call at all.

In this case, we add clamping after the inliner has run, so `i32s-rem` gets generated and called, instead of generated, used once, inlined, and removed.

Essentially: semantically these should all be the same, but the timing isn't identical and that perturbs the optimizer.

Reply:
Hm, interesting; it's actually not obvious whether it's better to clamp as early as possible (since it does actually seem better to just inline and remove `i32s-rem` if there's only one use), or later (since const-folding could eliminate calls before they get generated). Anyway we can land this one way and maybe experiment with that later.

Reply:
I currently run it as early as possible, which is necessarily after `AutoDrop` because the IR isn't valid before `AutoDrop` runs - asm2wasm seems to rely on AutoDrop cleaning it up, and we validate after each pass. We could run this separately in a separate PassRunner, or run AutoDrop and only AutoDrop separately in a separate PassRunner.

The other complication w/r/t timing in asm2wasm is the `OptimizingIncrementalModuleBuilder` that we use when running optimization passes that runs passes on each function - meaning the passes added to it need to be function-parallel, which we fundamentally can't make the clamping pass be because it adds functions to the module and thus introduces a sequence point for pass scheduling. Which we can probably work around with a separate pass runner. Maybe.

Reply:
Hmm, good points.

1. Running it after AutoDrop seems like the right thing, then. Not sure why we would need a separate pass runner? The issue, I think, is that we need to run it right after autodrop both when optimizing and when not, so there are two places to add to, as commented above (but I got the order in those two places wrong, should be after AutoDrop in both).
2. How about having a utility method that creates the helper functions, separate from the pass? That could be called in asm2wasm/s2wasm before creating any functions. Then the pass would only add calls to existing functions, and so would be a pure function-parallel pass. (The one downside is we might add some functions we don't need, but optimization would eliminate those.)

Reply:
The other downside to adding the functions before any optimizations, is that then in the pass we have to explicitly exclude any generated functions when running on functions, because otherwise we'll replace the `i32.div_u` instruction in the `$i32u-div` function with a call to `$i32u-div`.

The pass also gains a precondition, "must have called helper function `addTrappingFunctions` before running this pass".

These both make the pass less pass-like, but probably does a better job of maintaining asm2wasm's behavior exactly as it is today. I was operating under the assumption that that's a non-goal as long as the semantics are the same.

Reply:
I agree those two are downsides. But isn't the issue that we must do it that way to maintain the right semantics? Otherwise if the optimizer kicks in before we modify these math operations, it could optimize them incorrectly, as it's not seeing the correct math operations.

Otherwise we'd need to not run the optimizer until we run that pass, which would be substantially slower (we'd need to finish translating all functions before we start to optimize).

Another option might be to have some utility code that adds the functions, has code to do an individual fixup. And separately we'd have a pass that adds the functions and does a traversal to do all the fixups. So the pass could be 100% self-contained. And asm2wasm would not use that pass, it would call the utility code directly to do the fixups during translation (as it does now, but in a call to shared code), and also call the utility method to add the functions.

Reply:
Sounds very plausible. Let me try to add a test that exposes that behavior.

I originally implemented this as a pass for s2wasm and asm2wasm just calling utility functions directly, basically changing nothing about its flow (see https://github.com/WebAssembly/binaryen/pull/1168/commits/8ffe1c9ecdedba8df4564aecba0f60c204d4b313). I was able to use the pass for asm2wasm as well, which let me move the utility functions into the pass itself, which let me store the added functions on an instance variable on the pass object, instead of being a static variable, and generally make the interface much nicer.
Clean code does not trump *working* code, so I'll fix it.

Reply:
Added an `INT_MIN / -1` test in https://github.com/WebAssembly/binaryen/pull/1168/commits/691cf5e8af7c8f0e4c98b0f3a323336cdbc878d9. It does not currently misoptimize into a trap, and thinking about it a bit more I'm not sure if any optimization pass should ever do so?
Do these optimizations happen other than in `OptimizeInstructions.cpp`? Are there any patterns you can think of that would get optimized into an unconditional trap that I should check?

Reply:
Hmm you're right, currently Precompute will handle constants and branches, but not traps. But it could handle traps some day, although I guess that's low priority (and the type of trap would be lost, leaving a generic `(unreachable)`).

An issue currently affecting us would be that the side effects might be different (e.g. an FFI into JS should have more side effects than possibly trapping).

Anyhow, we don't need a specific new test for that I guess, seeing the current test outputs are identical should be enough.